### PR TITLE
New Add `vue/no-computed-in-data` Rule

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
Fixes: #958

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
finish https://github.com/vuejs/eslint-plugin-vue/issues/958

File: lib/utils/index.js [hunk 1]
```diff
@@ -464,7 +464,7 @@ module.exports = {
     return computedPropertiesNode.value.properties
       .filter(cp => cp.type === 'Property')
       .map(cp => {
-        const key = cp.key.name
+        const key = cp.key.type === 'Identifier' ? cp.key.name : cp.key.type === 'Literal' ? cp.key.value : undefined
         let value
```
Comment (IWANABETHATGUY) [inline lib/utils/index.js]: a property key can be a string literal, so i add this condition

**Comment by IWANABETHATGUY** [comment]
@ota-meshi could please help for review?

File: lib/rules/no-computed-in-data.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        if (node === dataAstNode) {
+          dataAstNode = undefined
+        }
+      },
+      MemberExpression (node) {
+        if (dataAstNode) {
+          // a memberExpression `like this.a.c.d` -> when traverse to c.d we can got the the full name -> this.a.c.d
+          const fullName = utils.parseMemberExpression(node).join('.')
+          if (memberExpressionMap[fullName]) {
+            memberExpressionMap[fullName] = [...memberExpressionMap[fullName], node]
```
Comment (ota-meshi) [inline lib/rules/no-computed-in-data.js]: Check the scope and change to trace correctly. `no-side-effects-in-computed-properties.js` is helpful.
Also add test cases to check if this works correctly.

```js
{
  data () {
    return {
      fn() { return this.foo }
    }
  },
  computed: {
    foo() {}
  }
}
```

File: lib/rules/no-computed-in-data.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+const utils = require('../utils')
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'require computed properties are not used in the data()',
+      categories: ['strongly-recommended'],
```
Comment (ota-meshi) [inline lib/rules/no-computed-in-data.js]: Could you change the categories to be set to undefined

File: lib/utils/index.js (hunk 1 above)
Comment (ota-meshi) [reply lib/utils/index.js]: Maybe you can use `getStaticPropertyName(cp)`.

File: lib/rules/no-computed-in-data.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      },
+      'Property:exit' (node) {
+        if (node === dataAstNode) {
+          dataAstNode = undefined
+        }
+      },
+      MemberExpression (node) {
+        if (dataAstNode) {
+          // a memberExpression `like this.a.c.d` -> when traverse to c.d we can got the the full name -> this.a.c.d
+          const fullName = utils.parseMemberExpression(node).join('.')
```
Comment (ota-meshi) [inline lib/rules/no-computed-in-data.js]: I think it's enough to check the first two.

```suggestion
          const fullName = utils.parseMemberExpression(node).slice(0,2).join('.')
```

I think this change eliminates the need to use startsWith. at 61.

File: lib/rules/no-computed-in-data.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          } else {
+            memberExpressionMap[fullName] = [node]
+          }
+        }
+      },
+      ...utils.executeOnVue(context, obj => {
+        const computedPropertyNameList = utils.getComputedProperties(obj).map(item => `this.${item.key}`)
+        Object.keys(memberExpressionMap).forEach(fullName => {
+          let index = -1
+          for (let i = 0, len = computedPropertyNameList.length; i < len; i++) {
```
Comment (ota-meshi) [inline lib/rules/no-computed-in-data.js]: Could you use findIndex?

File: docs/rules/no-computed-in-data.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+sidebarDepth: 0
+title: vue/no-computed-in-data
+description: require computed properties are not used in the data()
+---
+# vue/no-computed-in-data
+> require computed properties are not used in the data()
+
+- :gear: This rule is included in `"plugin:vue/strongly-recommended"` and `"plugin:vue/recommended"`.
+
+> require computed properties are not used in the data()
```
Comment (ota-meshi) [inline docs/rules/no-computed-in-data.md]: There is a duplicate description.

File: docs/rules/no-computed-in-data.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+- :gear: This rule is included in `"plugin:vue/strongly-recommended"` and `"plugin:vue/recommended"`.
+
+Please describe the origin of the rule here.
+
+## :book: Rule Details
+
+This rule report computed properties are used in data property
+
+Examples of **incorrect** code for this rule:
```
Comment (ota-meshi) [inline docs/rules/no-computed-in-data.md]: Could you rewrite a document like any other document?

`Examples of ** incorrect ** code for this rule:`
Is no longer used.

File: tests/lib/rules/no-computed-in-data.js [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+          }
+        },
+        computed: {
+          foo() {}
+        }
+      }
+      </script>
+      `
+    },
+    // should not warn when objectExpression is not a vue component
```
Comment (IWANABETHATGUY) [inline tests/lib/rules/no-computed-in-data.js:112]: @ota-meshi test for https://github.com/vuejs/eslint-plugin-vue/pull/1075/files#r400643684

File: lib/rules/no-computed-in-data.js (hunk 3 above)
Comment (IWANABETHATGUY) [reply lib/rules/no-computed-in-data.js:20]: image
if i set the categories to be undefined, the eslint will report this error

Comment (IWANABETHATGUY) [reply lib/rules/no-computed-in-data.js:20]: i change the categories same as `vue/no-side-effects-in-computed-properties`, is that ok?

Comment (IWANABETHATGUY) [reply lib/rules/no-computed-in-data.js]: i misunderstand your meaning before, i set it to undefined already😂

File: lib/rules/no-computed-in-data.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        }
+      },
+      ...utils.executeOnVue(context, obj => {
+        // check if targetObjectExpression is Vue component
+        if (targetObjectExpression !== obj) {
+          return
+        }
+        const computedPropertyNameList = utils.getComputedProperties(obj).map(item => `this.${item.key}`)
+        Object.keys(memberExpressionMap).forEach(fullName => {
+          const index = computedPropertyNameList.findIndex(name => fullName.startsWith(name))
```
Comment (ota-meshi) [inline lib/rules/no-computed-in-data.js:89]: There is no need to use startsWith.

```suggestion
          const index = computedPropertyNameList.findIndex(name => fullName === name)
```

File: lib/rules/no-computed-in-data.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      ':function:exit' (node) {
+        scopeStack.pop()
+      },
+      'Property:exit' (node) {
+        if (node === dataAstNode) {
+          dataAstNode = undefined
+        }
+      },
+      MemberExpression (node) {
+        if (dataAstNode && dataPropertyScope === topOfStack(scopeStack)) {
```
Comment (ota-meshi) [inline lib/rules/no-computed-in-data.js:68]: Vue components may not be at the top level.

```js
export default {
  components: {
    // @vue/component
    MyComponent: {
      data() {
        return {
          bar: this.baz
        }
      },
      computed: {
        baz() {}
      }
    }
  },
}
```

File: lib/rules/no-computed-in-data.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+          // a memberExpression `like this.a.c.d` -> when traverse to c.d we can got the the full name -> this.a.c.d
+          const fullName = utils.parseMemberExpression(node).slice(0, 2).join('.')
+          if (memberExpressionMap[fullName]) {
+            // check if parent node in this array, if true ignore this node, such as `{a: this.a.c.d}`, this traverse function visit order is `this.a.c.d` -> `a.c.d` -> `c.d`
+            const hasParentNodeInArray = memberExpressionMap[fullName].some(nodeInMap => node.parent === nodeInMap)
```
Comment (ota-meshi) [inline lib/rules/no-computed-in-data.js:73]: I've found an issue with duplicate reports when the chain is deep. Perhaps checking the `parent` is a problem.

```vue
      <script>
      export default {
        data() {
          return {
            a: this.test.foo.bar.baz,
          }
        },
        computed: {
          test() {
            return {}
          },
        }
      }
      </script>
```

File: lib/rules/no-computed-in-data.js [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+ * @author IWANABETHATGUY
+ */
+'use strict'
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+const utils = require('../utils')
+function topOfStack (stack) {
+  console.assert(Array.isArray(stack))
```
Comment (g-plane) [inline lib/rules/no-computed-in-data.js:12]: Why is there a `console.assert` call here?

**Comment by ota-meshi** [comment]
This PR seems to be out of date, so I will close it.

**Pull request closed without merging**